lib/fs.js

```diff
@@ -1666,7 +1668,8 @@ fs.realpathSync = function realpathSync(p, options) {
 
     // continue if not a symlink, break if a pipe/socket
     if (knownHard[base] || (cache && cache.get(base) === base)) {
-      if (isFileType(S_IFIFO) || isFileType(S_IFSOCK)) {
+      if (isFileType(statValues, S_IFIFO) ||
+          isFileType(statValues, S_IFSOCK)) {
```

Comment:
`statValues` is filled up by the `binding.lstat()` call ~20 lines up?  Is the logic here still correct if the entry comes from the cache?

Apropos that `lstat()` call, maybe do this?
```js
let stats;  // and use this in isFileType() calls
if (...) {
  const ctx = { path: base };
  stats = binding.lstat(pathModule.toNamespacedPath(base), undefined, ctx);
  // ...
}
```
(Also applies to `fs.realpath()`.)

Reply:
In the current code `statValues` always equals to `stats` since neither side allocate any new stats arrays. I could try this but from the code it's very hard to tell if stats will be `undefined`or not  when you hit the use statement below.

Reply:
Okay, never mind then.